### PR TITLE
Allow detached configuration to resolve current project

### PR DESCRIPTION
**Pull request by jvandort**
Previously, the root component of a detached configuration would claim to have the same gav as the project that created the detached configuration. This caused the resolution engine to treat the root component of the detached configuration as an instance of the project module.

However, the detached configuration's root component only has the detached configuration and none of the variants of the project. So, to resolve this we attach a suffix to the module ID of the root component of the detached configuration to ensure it does not conflict with the project.

This can be flimsy in certain situations. If there was another project with the same name as the constructed module ID (with the suffix) then the detached configuration's root component will conflict with that component. However, this is likely rare, as the constructed module ID would look something like org:module-detachedConfiguration1:version or org:module-runtimeClasspathCopy:version.

Future work could aim to improve this by adding a ComponentIdentifier-like counterpart to ModuleIdentifier. Then, we could allow the root component here to have some opaque module identifier that does not collide with any named module identifier

Partially affects https://github.com/gradle/gradle/issues/27947
Fixes https://github.com/gradle/gradle/issues/24126

Potentially related https://github.com/gradle/gradle/issues/15032

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

**Comment by jvandort** [comment]
@bot-gradle test this

**Comment by jvandort** [comment]
@bot-gradle test this

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/moduleconverter/RootComponentMetadataBuilder.java [hunk 1]
```diff
@@ -15,21 +15,46 @@
...
     RootComponentState toRootComponent(String configurationName);
 
-    RootComponentMetadataBuilder withConfigurationsProvider(ConfigurationsProvider provider);
+    /**
+     * Create a new builder, that builds a new component with a new identity and configuration set
+     */
+    RootComponentMetadataBuilder newBuilder(DependencyMetaDataProvider identity, ConfigurationsProvider provider);
+
+    /**
+     * Get the identity of the component build by this builder.
```
Comment (jvandort) [inline platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/moduleconverter/RootComponentMetadataBuilder.java:42]:
```suggestion
     * Get the identity of the component built by this builder.
```

**Comment by jvandort** [comment]
@bot-gradle test this

**Review by octylFractal — APPROVED** [review]

**Pull request merged**